### ISSUE TITLE
fread memory leak when small number bumped up to character

### ISSUE THREAD

**Issue by alexdeng** [Rdatatable/data.table#918]
I have a 300m file with some small numbers in scientific notation (E-300) and after using fread it used more than 30gig of my memory. After manually use colClass to specify the column to be character the problem is gone and the memory usage is normal. I suspect there is memory leak somewhere when handling small number.

**Comment by arunsrinivasan** [comment]
Yes, I can confirm the steady increase of memory on Windows.

**Comment by braidm** [comment]
I'd love to see this fixed. Reading my 30MB file which contains some super small numbers leaves me with 5+ GB of leaked memory on Windows, causing me to have to restart between executions. I could treat as character then coerce, but what a pain.

**Comment by st-pasha** [comment]
@braidm Can you provide some more context please? How small are "super small numbers"? Say, can you post few lines of your file as an example?
The OP's test file is no longer accessible...

**Comment by braidm** [comment]
I cannot share my original file which has 160,000 rows, 28 fields, tab delimited, and leaks over 7GB whenever I read it (on disk it is 30MB). The leak does not happen if all numerics are larger than the machine limit of 2.225074e-308.

However I created a test file which leaks about 300MB each time it is read. With repeated executions it will leak multiple GB. Here is the message from data.table

`> leak <- fread("data/leak.txt",header=TRUE,sep="\t",quote="",stringsAsFactors = FALSE,select=c("pvalue"))
Warning message:
In fread("data/leak.txt", header = TRUE, sep = "\t", quote = "",  :
  C function strtod() returned ERANGE for one or more fields. The first was string input '0.58E-2141'. It was read using (double)strtold() as numeric value 0.0000000000000000E+00 (displayed here using %.16E); loss of accuracy likely occurred. This message is designed to tell you exactly what has been done by fread's C code, so you can search yourself online for many references about double precision accuracy and these specific C functions. You may wish to use colClasses to read the column as character instead and then coerce that column using the Rmpfr package for greater accuracy.`

[leak.txt](https://github.com/Rdatatable/data.table/files/1134358/leak.txt)

and here is the sessionInfo
`> sessionInfo()
R version 3.3.2 (2016-10-31)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.10.4

loaded via a namespace (and not attached):
[1] tools_3.3.2`

**Comment by st-pasha** [comment]
@braidm Thanks for providing the test data. I was able to verify that the "leak" occurs on a Windows machine with data.table `1.10.4`. Well, actually I found that when I ran `fread("leak.txt")` in a loop, the process eats the memory at an enormous speed, quickly reaching 100% RAM utilization, at which point everything comes to a crawling speed. However, after the loop ends, the memory will slowly be reclaimed, eventually reaching the normal level. I suspect what happens is that on Windows the warning messages are somehow not being processed at the right time, which results in this observed behavior.

Notably, the same problem does not occur on a Linux machine, or a MacOS machine. The good news is that the problem doesn't occur with latest development version of `data.table` on Windows either (simply because these ERANGE warnings are not emitted). You can install the latest version via
```
remove.packages("data.table")
install.packages("https://ci.appveyor.com/api/buildjobs/e27ehri93y0tkj6f/artifacts/data.table_1.10.5.zip", repos=NULL)
```
Please let me know if this solves your problem

**Comment by braidm** [comment]
I'm grateful for you taking a look. I cannot try the dev version of data.tables at this time. You seem to be suggesting the memory is not truly leaked, because it does and can get garbage collected. What is strange then is that in RStudio when the memory has ballooned up to 8GB after reading a 30MB file, that a subsequent run pushes the memory demand yet higher into the teens of GB when in fact if it only did gc we could be OK. And if it's not a gc issue, how else would the memory be reclaimed. Thanks again!

**Comment by mattdowle** [comment]
Talking to Pasha we've realized where the 'leak' arises. Each and every warning is slightly different because it includes the string value in this part of the warning message: `"The first was string input '0.58E-2141'."`  R adds each unique string to its global string cache regardless of what each string is needed for.  So R's global string cache is growing with the huge number of warning messages being constructed here.

dev is much better and the problem has gone away.

Just a test needed then to be added to the test suite, and this is closed.

**Comment by MichaelChirico** [comment]
Closed by #2451
